src/devices/src/virtio/net/device.rs

```diff
@@ -497,242 +508,102 @@ impl NetEpollHandler {
-                // Process a deferred frame first if available. Don't read from tap again
-                // until we manage to receive this deferred frame.
-                {
-                    if self.rate_limited_rx_single_frame() {
-                        self.rx.deferred_frame = false;
-                        self.process_rx()
-                    } else if self.rx.deferred_irqs {
-                        self.rx.deferred_irqs = false;
-                        self.signal_used_queue()
-                    } else {
-                        Ok(())
-                    }
-                } else {
-                    self.process_rx()
-                }
-            }
-            TX_QUEUE_EVENT => {
-                METRICS.net.tx_queue_event_count.inc();
-                if let Err(e) = self.tx.queue_evt.read() {
-                    error!("Failed to get tx queue event: {:?}", e);
-                    METRICS.net.event_fails.inc();
-                    Err(DeviceError::FailedReadingQueue {
-                        event_type: "tx queue event",
-                        underlying: e,
-                    })
-                } else if !self.tx.rate_limiter.is_blocked()
-                // If the limiter is not blocked, continue transmitting bytes.
-                {
-                    self.process_tx()
-                } else {
-                    Ok(())
-                }
-            }
-            RX_RATE_LIMITER_EVENT => {
-                METRICS.net.rx_event_rate_limiter_count.inc();
-                // Upon rate limiter event, call the rate limiter handler
-                // and restart processing the queue.
-                match self.rx.rate_limiter.event_handler() {
-                    Ok(_) => {
-                        // There might be enough budget now to receive the frame.
-                        self.resume_rx()
-                    }
-                    Err(e) => {
-                        METRICS.net.event_fails.inc();
-                        error!("Failed to get rx rate-limiter event: {:?}", e);
-                        Err(DeviceError::RateLimited(e))
-                    }
-                }
-            }
-            TX_RATE_LIMITER_EVENT => {
-                METRICS.net.tx_rate_limiter_event_count.inc();
-                // Upon rate limiter event, call the rate limiter handler
-                // and restart processing the queue.
-                match self.tx.rate_limiter.event_handler() {
-                    Ok(_) => {
-                        // There might be enough budget now to send the frame.
-                        self.process_tx()
-                    }
-                    Err(e) => {
-                        METRICS.net.event_fails.inc();
-                        error!("Failed to get tx rate-limiter event: {:?}", e);
-                        Err(DeviceError::RateLimited(e))
-                    }
-                }
+        if let Err(e) = self.queue_evts[RX_INDEX].read() {
```

Comment:
Nit: I am unsure how the compiler/llvm optimizes this, but it looks like this path favors the error cases in terms of branching. I would test for Ok and then for Err if that makes sense.